### DISCUSSION TITLE
Why is my neural network not converging for a 4D system of uncoupled pendulums?

### DISCUSSION THREAD

**Discussion by motavini** [lululxvi/deepxde#1238] [Q&A]
Hi everyone,

I'm currently working on a project that involves solving a system of uncoupled pendulums. 

The input to the network is a 4D vector consisting of the angle and angular velocity for each pendulum. 

However, I'm having trouble getting the network to converge. The loss function decreases initially but appears to plateau after around 1000 iterations, with little to no improvement after that. I'm using the Adam optimizer with a learning rate of 0.001.

I suspect that the problem may be due to the 4-dimensional nature of the problem. Is this a common issue with higher-dimensional systems, or am I missing something else? Any suggestions or insights would be greatly appreciated.

Here is my code:

```
import deepxde as dde
import numpy as np
import math
import tensorflow as tf

g = 9.81
L1 = 1
L2 = 1
theta1 = np.pi/18
theta1d = 0.0
theta2 = np.pi/36
theta2d = 0.0
def ode_system(x, y):
    
    y1, y2, y3, y4 = y[:, 0], y[:, 1], y[:, 2], y[:, 3]
    dy1_x = dde.grad.jacobian(y, x, i=0)
    dy2_x = dde.grad.jacobian(y, x, i=1)
    dy3_x = dde.grad.jacobian(y, x, i=2)
    dy4_x = dde.grad.jacobian(y, x, i=3)
    eq1 = -(g/L1)*y1
    eq2 = -(g/L2)*y3
    return [dy1_x - y2, dy2_x - eq1, dy3_x - y4, dy4_x - eq2]


def boundary(_, on_initial):
    return on_initial


def func(x):

    return np.hstack((np.pi*np.cos(math.sqrt(g/L1)*x)/18, np.pi*np.cos(math.sqrt(g/L2)*x)/36 ,-math.sqrt(g/L1)*np.pi*np.sin(math.sqrt(g/L1)*x)/18,-math.sqrt(g/L2)*np.pi*np.sin(math.sqrt(g/L2)*x)/36))


geom = dde.geometry.TimeDomain(0, 2)
ic1 = dde.icbc.IC(geom, lambda x: theta1, boundary, component=0)
ic2 = dde.icbc.IC(geom, lambda x: theta1d, boundary, component=1)
ic3 = dde.icbc.IC(geom, lambda x: theta2, boundary, component=2)
ic4 = dde.icbc.IC(geom, lambda x: theta2d, boundary, component=3)
data = dde.data.PDE(geom, ode_system, [ic1, ic2, ic3, ic4], 400, 4, solution=func, num_test=100)

layer_size = [1] + [50] * 3 + [4]
activation = "tanh"
initializer = "Glorot uniform"
net = dde.nn.FNN(layer_size, activation, initializer)

def input_transform(t):
    return tf.concat(
        (
            t,
            tf.sin(t),
            tf.sin(2 * t),
            tf.sin(3 * t),
            tf.sin(4 * t),
            tf.sin(5 * t),
            tf.sin(6 * t),
            tf.sin(7 * t),
            tf.sin(8 * t),
            tf.sin(9 * t),
        ),
        axis=1,
    )
net.apply_feature_transform(input_transform)

model = dde.Model(data, net)
model.compile("adam", lr=0.001)
losshistory, train_state = model.train(iterations=30000)
model.compile("L-BFGS", lr=0.001)
losshistory, train_state = model.train()

dde.saveplot(losshistory, train_state, issave=True, isplot=True)

```

Here is what I get reported back
```
Step      Train loss                                                                          Test loss                                                                           Test metric
0         [6.03e+00, 2.70e+01, 2.82e+00, 1.15e+01, 3.05e-02, 0.00e+00, 7.62e-03, 0.00e+00]    [5.99e+00, 2.63e+01, 2.78e+00, 1.14e+01, 3.05e-02, 0.00e+00, 7.62e-03, 0.00e+00]    []  
1000      [1.63e-03, 2.41e-03, 4.34e-04, 6.11e-04, 2.16e-02, 3.54e-05, 5.40e-03, 8.86e-06]    [1.42e-03, 3.59e-03, 3.68e-04, 9.05e-04, 2.16e-02, 3.54e-05, 5.40e-03, 8.86e-06]    []  
2000      [1.62e-03, 2.40e-03, 4.19e-04, 6.09e-04, 2.16e-02, 3.53e-05, 5.40e-03, 8.72e-06]    [1.39e-03, 3.58e-03, 3.50e-04, 9.04e-04, 2.16e-02, 3.53e-05, 5.40e-03, 8.72e-06]    []  
3000      [1.62e-03, 2.40e-03, 4.09e-04, 6.03e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.76e-06]    [1.39e-03, 3.58e-03, 3.45e-04, 8.98e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.76e-06]    []  
4000      [1.59e-03, 2.47e-03, 3.66e-04, 8.29e-04, 2.17e-02, 3.79e-05, 5.39e-03, 7.74e-06]    [1.37e-03, 3.62e-03, 3.22e-04, 1.13e-03, 2.17e-02, 3.79e-05, 5.39e-03, 7.74e-06]    []  
5000      [1.61e-03, 2.40e-03, 4.02e-04, 6.00e-04, 2.16e-02, 3.52e-05, 5.41e-03, 8.82e-06]    [1.39e-03, 3.58e-03, 3.46e-04, 8.94e-04, 2.16e-02, 3.52e-05, 5.41e-03, 8.82e-06]    []  
6000      [2.03e-03, 2.55e-03, 4.59e-04, 6.36e-04, 2.24e-02, 3.57e-05, 5.54e-03, 6.39e-06]    [1.76e-03, 3.45e-03, 3.96e-04, 8.70e-04, 2.24e-02, 3.57e-05, 5.54e-03, 6.39e-06]    []  
7000      [1.47e-03, 3.27e-03, 4.16e-04, 6.10e-04, 2.13e-02, 3.79e-05, 5.43e-03, 1.05e-05]    [1.27e-03, 4.58e-03, 3.59e-04, 8.94e-04, 2.13e-02, 3.79e-05, 5.43e-03, 1.05e-05]    []  
8000      [1.66e-03, 2.42e-03, 3.72e-04, 6.08e-04, 2.16e-02, 3.43e-05, 5.46e-03, 6.75e-06]    [1.36e-03, 3.61e-03, 3.26e-04, 8.89e-04, 2.16e-02, 3.43e-05, 5.46e-03, 6.75e-06]    []  
9000      [1.62e-03, 2.40e-03, 4.03e-04, 6.04e-04, 2.16e-02, 3.49e-05, 5.40e-03, 8.85e-06]    [1.36e-03, 3.58e-03, 3.45e-04, 8.99e-04, 2.16e-02, 3.49e-05, 5.40e-03, 8.85e-06]    []  
10000     [1.62e-03, 2.39e-03, 4.08e-04, 6.00e-04, 2.16e-02, 3.49e-05, 5.40e-03, 8.78e-06]    [1.36e-03, 3.57e-03, 3.49e-04, 8.96e-04, 2.16e-02, 3.49e-05, 5.40e-03, 8.78e-06]    []  
11000     [1.62e-03, 2.39e-03, 4.05e-04, 6.00e-04, 2.16e-02, 3.50e-05, 5.40e-03, 8.80e-06]    [1.36e-03, 3.57e-03, 3.46e-04, 8.95e-04, 2.16e-02, 3.50e-05, 5.40e-03, 8.80e-06]    []  
12000     [1.62e-03, 2.39e-03, 4.03e-04, 6.01e-04, 2.16e-02, 3.52e-05, 5.41e-03, 8.78e-06]    [1.37e-03, 3.57e-03, 3.44e-04, 8.95e-04, 2.16e-02, 3.52e-05, 5.41e-03, 8.78e-06]    []  
13000     [1.61e-03, 2.39e-03, 4.01e-04, 5.99e-04, 2.16e-02, 3.53e-05, 5.41e-03, 8.77e-06]    [1.37e-03, 3.57e-03, 3.43e-04, 8.94e-04, 2.16e-02, 3.53e-05, 5.41e-03, 8.77e-06]    []  
14000     [1.62e-03, 2.39e-03, 4.04e-04, 5.98e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.80e-06]    [1.37e-03, 3.57e-03, 3.44e-04, 8.94e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.80e-06]    []  
15000     [1.63e-03, 2.37e-03, 4.05e-04, 6.00e-04, 2.16e-02, 3.51e-05, 5.40e-03, 8.98e-06]    [1.39e-03, 3.55e-03, 3.46e-04, 8.97e-04, 2.16e-02, 3.51e-05, 5.40e-03, 8.98e-06]    []  
16000     [1.61e-03, 2.40e-03, 4.04e-04, 6.00e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.79e-06]    [1.38e-03, 3.58e-03, 3.46e-04, 8.96e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.79e-06]    []  
17000     [1.62e-03, 2.39e-03, 4.04e-04, 5.99e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.79e-06]    [1.38e-03, 3.58e-03, 3.47e-04, 8.95e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.79e-06]    []  
18000     [1.59e-03, 2.50e-03, 4.09e-04, 6.13e-04, 2.16e-02, 2.95e-05, 5.38e-03, 9.40e-06]    [1.36e-03, 3.71e-03, 3.53e-04, 9.16e-04, 2.16e-02, 2.95e-05, 5.38e-03, 9.40e-06]    []  
19000     [1.58e-03, 2.41e-03, 3.94e-04, 6.04e-04, 2.16e-02, 3.49e-05, 5.41e-03, 8.66e-06]    [1.38e-03, 3.58e-03, 3.41e-04, 8.96e-04, 2.16e-02, 3.49e-05, 5.41e-03, 8.66e-06]    []  
20000     [1.61e-03, 2.40e-03, 4.03e-04, 6.00e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.81e-06]    [1.39e-03, 3.58e-03, 3.47e-04, 8.95e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.81e-06]    []  
21000     [1.61e-03, 2.40e-03, 4.04e-04, 5.99e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.80e-06]    [1.39e-03, 3.58e-03, 3.47e-04, 8.95e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.80e-06]    []  
22000     [1.62e-03, 2.39e-03, 4.04e-04, 5.99e-04, 2.16e-02, 3.51e-05, 5.40e-03, 8.78e-06]    [1.39e-03, 3.58e-03, 3.47e-04, 8.94e-04, 2.16e-02, 3.51e-05, 5.40e-03, 8.78e-06]    []  
23000     [1.62e-03, 2.39e-03, 4.04e-04, 5.99e-04, 2.16e-02, 3.51e-05, 5.40e-03, 8.78e-06]    [1.39e-03, 3.58e-03, 3.47e-04, 8.95e-04, 2.16e-02, 3.51e-05, 5.40e-03, 8.78e-06]    []  
24000     [1.62e-03, 2.39e-03, 4.04e-04, 5.99e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.78e-06]    [1.39e-03, 3.58e-03, 3.48e-04, 8.95e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.78e-06]    []  
25000     [1.59e-03, 2.25e-03, 4.36e-04, 1.27e-03, 2.19e-02, 3.73e-05, 5.13e-03, 1.09e-05]    [1.37e-03, 3.33e-03, 3.79e-04, 1.67e-03, 2.19e-02, 3.73e-05, 5.13e-03, 1.09e-05]    []  
26000     [1.62e-03, 2.39e-03, 4.03e-04, 6.13e-04, 2.16e-02, 3.51e-05, 5.39e-03, 8.41e-06]    [1.39e-03, 3.58e-03, 3.50e-04, 9.13e-04, 2.16e-02, 3.51e-05, 5.39e-03, 8.41e-06]    []  
27000     [1.62e-03, 2.42e-03, 4.03e-04, 6.10e-04, 2.16e-02, 3.62e-05, 5.39e-03, 8.29e-06]    [1.41e-03, 3.61e-03, 3.54e-04, 9.09e-04, 2.16e-02, 3.62e-05, 5.39e-03, 8.29e-06]    []  
28000     [1.61e-03, 2.40e-03, 4.09e-04, 5.99e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.77e-06]    [1.41e-03, 3.58e-03, 3.42e-04, 8.95e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.77e-06]    []  
29000     [1.62e-03, 2.40e-03, 4.06e-04, 5.99e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.76e-06]    [1.41e-03, 3.58e-03, 3.43e-04, 8.94e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.76e-06]    []  
30000     [1.62e-03, 2.40e-03, 4.05e-04, 5.99e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.76e-06]    [1.41e-03, 3.58e-03, 3.45e-04, 8.95e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.76e-06]    []  

Best model at step 27000:
  train loss: 3.21e-02
  test loss: 3.33e-02
  test metric: []

'train' took 465.987356 s

Compiling model...
Warning: learning rate is ignored for L-BFGS
'compile' took 1.481581 s

Training model...

Step      Train loss                                                                          Test loss                                                                           Test metric
30000     [1.62e-03, 2.40e-03, 4.05e-04, 5.99e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.76e-06]    [1.41e-03, 3.58e-03, 3.45e-04, 8.95e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.76e-06]    []  
30026     [1.62e-03, 2.40e-03, 4.06e-04, 5.98e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.76e-06]    [1.41e-03, 3.58e-03, 3.46e-04, 8.94e-04, 2.16e-02, 3.52e-05, 5.40e-03, 8.76e-06]    []  

```

Thank you in advance
